Import in Signup: Validate site is importable

creativecoder opened the pull request:
## Summary

When importing a sign during signup, validates the url and checks the importability of the site against `is-site-importable` endpoint. Shows an appropriate error message if the url is incorrect or the site is not importable.

## Tasks

- [x] Input validation UI
- [x] Check site is importable before proceeding to the next step
- [x] Loading state while checking site import-ability
- [x] Styles for notice positioning, card margins, etc
- [ ] [punted until endpoint is updated]`is-site-importable` error message that differentiates between not finding the site and the site not being compatible with importer
- [x] Importer specific url validation (from `client/my-sites/importer/site-importer/site-importer-input-pane.jsx`)

## Testing Plan

1. Visit `/start/import/from-url`
1. Try submitting with a blank url; you should see an error.
1. Try submitting with an invalid url; you should see an error.
1. Try submitting with a valid url that isn't a real site; you should see a loading state as the site is checked, then an error.
1. Try submitting with a url without a path (i.e. `https://me.wixsite.com`); you should see an error.
1. Try submitting with an admin url (i.e. `https://www.wix.com/website/builder`); you should see an error.
1. Try submitting with an actual site url that is importable; you should be able to proceed to the next step.

creativecoder commented on client/my-sites/importer/site-importer/url-validation.js:
```diff
@@ -0,0 +1,45 @@
+ */
+import url from 'url';
+import { isWebUri } from 'valid-url';
+import { translate } from 'i18n-calypso';
```
Is this a valid way to use translate function (outside of React components)?

spen replied:
Sure thing!
IIRC, the only gotcha with using translate directly is when it's called before the translations are properly loaded - in which case I think the base string would be returned.

This might happen if, say, we called translate in the files global scope, outside of functions etc:

```js
import { translate } from 'i18n-calypso';
import { inRange, trim } from 'lodash';

const someMessage = translate( 'some message' );
```

But because here we're calling translate from a function that'll be called by a component, we can safely assume that the translations are loaded at the time of calling `translate` 👌

creativecoder replied:
👍

creativecoder commented on client/state/importer-nux/actions.js:
```diff
@@ -6,7 +6,6 @@ import {
 	IMPORT_IS_SITE_IMPORTABLE_ERROR,
 	IMPORT_IS_SITE_IMPORTABLE_RECEIVE,
 	IMPORTER_NUX_URL_INPUT_SET,
-	IMPORTER_NUX_URL_VALIDATION_SET,
```
Remove url validation message from global state since it isn't shared outside component.

creativecoder commented:
@spen @jblz This still needs styling work, but should be ready for an initial review.

spen commented on client/signup/steps/import-url/index.jsx:
```diff
@@ -5,162 +5,211 @@
 	render() {
-		const { flowName, positionInFlow, signupProgress, stepName, translate } = this.props;
+		const { flowName, isLoading, positionInFlow, signupProgress, stepName, translate } = this.props;
+		const noticeText = translate( "Please wait, we're checking this URL." );
```
Would it be overkill to mention that were checking whether or not the URL is importable? - that

It might actually be better to use the term 'site' too as that's a bit less technical.

creativecoder replied:
I like that idea, will modify.

spen commented on client/my-sites/importer/site-importer/url-validation.js:
```diff
@@ -0,0 +1,45 @@
+	return isWebUri( url.format( parsedUrl ) ) && hasTld;
+};
+
+export function validateImportUrl( value ) {
```
Shall we cut out the original source of this validation from `site-importer-input-pane.jsx` and put this utility to work there as part of this PR? Or shall we save that for a follow up? I'd be happy with either :)

creativecoder replied:
I vote for any changes to Settings > Import to happen in a follow-up to this PR, as I have a sense there may be some other clean-up and improvements that can happen at the same time.

creativecoder replied:
I went ahead and updated so that this is used in `client/my-sites/importer/site-importer/site-importer-input-pane.jsx` as well.

spen commented on client/signup/steps/import-url/index.jsx:
```diff
@@ -5,162 +5,211 @@
+			! isEqual( prevProps.isSiteImportableError, isSiteImportableError ) &&
+			isSiteImportableError
+		) {
+			this.focusInput();
```
Nice addition 👌

spen commented on client/signup/steps/import-url/index.jsx:
```diff
@@ -5,162 +5,211 @@
-	// @TODO sanitize? Prepend https:// ..?
-	return url;
-};
+// const normalizeUrlForImportSource = url => {
```
Lets chop this out, if necessary, before we merge 👍

creativecoder replied:
Will do.

creativecoder commented on client/my-sites/importer/site-importer/url-validation.js:
```diff
@@ -0,0 +1,45 @@
+		);
+	} else if ( hostname.indexOf( '.wixsite.com' ) > -1 && pathname === '/' ) {
+		return translate(
+			"You haven't entered the full URL. Please include the part of the URL that comes after wixsite.com. See below for an example."
```
Agreed--thanks for flagging. I'll post a separate comment about that in the main thread of this PR.

creativecoder replied:
Example urls have been added to the signup flow page.

spen replied:
Awesome, cheers @creativecoder !

creativecoder commented on client/state/importer-nux/reducer.js:
```diff
@@ -32,19 +28,19 @@ export const siteDetails = createReducer(
-	[ IMPORT_IS_SITE_IMPORTABLE_ERROR ]: ( state, action ) => action.error,
+	[ IMPORT_IS_SITE_IMPORTABLE_START_FETCH ]: () => null,
+	[ IMPORT_IS_SITE_IMPORTABLE_RECEIVE ]: () => null,
+	[ IMPORT_IS_SITE_IMPORTABLE_ERROR ]: ( _state, action ) => action.error,
```
Yes, prefixing with an underscore is a convention to indicate the parameter is unused. Typescript has adopted this as a way to tell the compiler you don't want to be warned about not using a parameter and it's used in ReasonML as well.

If this isn't done anywhere else in Calypso, I'd rather take it out to be consistent.

a8ci18n commented on client/signup/steps/import-url/index.jsx:
```diff
@@ -2,132 +2,197 @@
+
+				<div className="import-url__example">
+					<p className="import-url__example-urls">
+						{ translate( 'Example URLs', {
```
:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

creativecoder commented on client/components/domains/example-domain-browser/index.jsx:
```diff
@@ -0,0 +1,32 @@
+/** @format */
```
Move example browser svg to its own component so it can be reused in the import url step.

creativecoder commented on client/my-sites/importer/site-importer/site-importer-input-pane.jsx:
```diff
@@ -184,19 +185,7 @@ class SiteImporterInputPane extends React.Component {
-				"You haven't entered the full URL. Please include the part of the URL that comes after wixsite.com. See below for an example."
-			);
-		}
+		const errorMessage = validateImportUrl( siteURL );
```
Move url validation logic into a separate file so it can be shared between Settings > Import and the import signup flow.

jancavan commented:
Nice work, @creativecoder!

- When it's invalid, we use a red border as a standard:
![screenshot 2018-10-10 22 28 34](https://user-images.githubusercontent.com/4924246/46782700-04312080-ccdc-11e8-9147-bf6b0934710b.png)
- When I click on the field and click outside of it, it validates it immediately and tells me to "Please enter a valid URL"
- Sorry for this change. It's probably better for the placeholder field value to be reverted to "Please enter the full URL of your site" since we now have the graphic to explain what it is.

It does create a jumpiness when there's an error. I'll fix that separately.

jancavan commented on client/components/domains/example-domain-browser/style.scss:
```diff
@@ -0,0 +1,28 @@
+		display: block;
+		content: '';
+		position: absolute;
+		right: 0;
```
Position properties need to be indented after position declaration.

jancavan commented on client/components/domains/example-domain-browser/style.scss:
```diff
@@ -0,0 +1,28 @@
+	&::before {
+		height: 30px;
+		background: linear-gradient(
+			to bottom,
```
This can be in a single line.

jancavan commented on client/components/domains/example-domain-suggestions/style.scss:
```diff
@@ -28,37 +30,14 @@
 	position: absolute;
-		top: 0;
-		right: 0;
+	top: 0;
```
Also indent.

jancavan commented on client/signup/steps/import-url/style.scss:
```diff
@@ -0,0 +1,76 @@
+
+	&__notice {
+		position: absolute;
+		top: -88px;
```
Also nested.

jblz commented:
This is working really well and the code looks good beyond what's been mentioned.

Nice work!

---

I have a non-blocking UX suggestion (feel free to push to an issue or follow up PR):

If I end up here by mistake, this message is useful in that it tells me I _may_ still be able to import my content, but only by allusion:

<img width="928" alt="screen shot 2018-10-11 at 10 40 55 am" src="https://user-images.githubusercontent.com/1587282/46812303-70df0600-cd42-11e8-8e01-47ea3f64b714.png">

IMO We should link to:
* `/start` (can be a `Start Over` button or similar in the UI)
* https://support.wordpress.com/import

...& suggest they try again once they create an account.

jancavan commented:
@creativecoder A couple other comments:

- Not sure why the global notice is smaller than normal? Should also be flushed right.
![screenshot 2018-10-11 10 04 47](https://user-images.githubusercontent.com/4924246/46821367-49d20580-cd3d-11e8-9bdc-08735d7cf44d.png)

- As the URL is being checked, can the wait button change to "Checking..."?
![screenshot 2018-10-11 10 05 09](https://user-images.githubusercontent.com/4924246/46821435-7ede5800-cd3d-11e8-95bb-bedfd44d5fb2.png)

spen commented:
I had another run through the code and it looks great :) GW!
It looks like the `ci/wp-desktop` tests are failing, but I can't figure out why that is. I tried to trigger another test but we still get the same issue: 

> fatal: Couldn't find remote ref tests/add/signup-import-landing-handler--validate-importable-site

creativecoder commented:
Created https://github.com/Automattic/delta-private/issues/71 (which needs a back-end importer change) to follow up on this task

> is-site-importable error message that differentiates between not finding the site and the site not being compatible with importer

jancavan commented:
Discussed during the call, but just adding it again here for completeness. 

As mentioned in: https://github.com/Automattic/delta-private/issues/32#issuecomment-429194056, let's add an escape route for users who end up in Step 1 by mistake:

![screenshot 2018-10-12 08 34 59](https://user-images.githubusercontent.com/4924246/46882580-055d6d00-ce04-11e8-9cfd-95aee9b07f13.png)

- "Other sources" links to https://en.support.wordpress.com/import/ (target `_blank`)
- "Sign up" links to the regular signup flow. Was initially thinking of dropping them off onto the `/import` page, but let's just give them Checklist for now.

Can we make sure to add tracking to these links please? Thanks!

creativecoder commented:
PR has been updated

1. UI updates based on previous comments
1. Add "escape route" at the bottom of the step, in case user doesn't have a Wix site
1. Added tracks events `calypso_signup_exit_flow` and `calypso_signup_support_clicked` for "escape route" actions

@jancavan Ready for another design review. In addition to adding the default "Please enter the full URL..." message, I changed the url input validation trigger so the validation message doesn't update until blur--let me know if there's a different behavior you think would be better.

jancavan commented:
@creativecoder Small nitpick: When I click on "Sign up", the entire text moves down with the button as seen in: https://cloudup.com/cZId4iovzq1

jblz commented on client/components/domains/example-domain-suggestions/style.scss:
```diff
@@ -16,7 +18,7 @@
 	}
 
-	@include breakpoint( "<660px" ) {
+	@include breakpoint( '<660px' ) {
```
Why is this scss file changing double quotes to single quotes while the others are doing the opposite?

creativecoder replied:
This was me fighting with prettier formatting .scss files--converted back to `"`, which seems to be what's usually used.

jblz replied:
It's so helpful until it's not :)

jblz commented on client/my-sites/importer/site-importer/url-validation.js:
```diff
@@ -0,0 +1,47 @@
+import { inRange, trim } from 'lodash';
+
+const parseUrl = function( value = '' ) {
+	const parsedUrl = url.parse( value, false, true );
```
Do we need to wrap the `url.parse` in a try / catch? It throws `TypeError` & `URIError`.

jblz commented on client/signup/steps/import-url/index.jsx:
```diff
@@ -2,165 +2,279 @@
-	getUrlInputValidationMessage,
 	isUrlInputDisabled,
 } from 'state/importer-nux/temp-selectors';
+import { validateImportUrl } from '../../../my-sites/importer/site-importer/url-validation';
```
It's probably better to do:
`import { validateImportUrl } from 'my-sites/importer/site-importer/url-validation';`

...instead of so many `../`s

creativecoder replied:
Decided to move this to `lib/importers` and updated path.

jblz commented on client/signup/steps/import-url/index.jsx:
```diff
@@ -2,165 +2,279 @@
+	getUrlMessage = () => {
+		if ( this.state.urlValidationMessage ) {
+			return this.state.urlValidationMessage;
+		} else if ( this.props.isSiteImportableError ) {
```
Don't need the `else if` when returning from the above `if` block.  Just use a new `if`

creativecoder replied:
Fixed.

jblz commented on client/signup/steps/import-url/index.jsx:
```diff
@@ -2,165 +2,279 @@
+		const { isSiteImportableError, translate } = this.props;
+
+		switch ( isSiteImportableError.code ) {
+			case 1000001:
```
These should be enumerated constants that this component imports.

creativecoder replied:
Excellent idea. Updated.

jblz commented on client/signup/steps/import-url/index.jsx:
```diff
@@ -2,165 +2,279 @@
+							busy={ isLoading }
+							type="submit"
+						>
+							{ isLoading ? translate( 'Checking' ) + '...' : translate( 'Continue' ) }
```
This could have unexpected results in RTL languages:
`translate( 'Checking' ) + '...'`

Is there any reason you didn't just include an ellipsis in the string to translate?

Also, probably better to use an ellipses character `…` instead of three dots.

creativecoder replied:
I started with `&hellip;` but struggled with it being escaped by `translate`. Got it figured out now.

jblz commented on client/signup/steps/import-url/index.jsx:
```diff
@@ -2,165 +2,279 @@
 	render() {
-		const { flowName, positionInFlow, signupProgress, stepName, translate } = this.props;
+		const { flowName, isLoading, positionInFlow, signupProgress, stepName, translate } = this.props;
+		const noticeText = translate(
```
Any reason to assign this rather than just setting the attribute value?

akirk commented on client/lib/importers/url-validation.js:
```diff
@@ -0,0 +1,47 @@
+		);
+	} else if ( hostname.indexOf( '.wixsite.com' ) > -1 && pathname === '/' ) {
+		return translate(
+			'Please enter the full URL and include the part that comes after wixsite.com. See example below.'
```
Could we make `wixsite.com` a variable so that translators don't accidentially change it?

akirk replied:
Nevermind, this string is already properly translated and you just moved it.

jancavan commented:
@creativecoder Sorry I didn't catch this earlier, but can we change the value "https://example.wixsite.com/mysite" to "account.wixsite.com/my-site", just to make it the same as the example?

creativecoder commented:
Changes so far today:

@jancavan 

- Thanks for the style tweaks!
- I made url input placeholder and example match (both include "https://", because I think that's more immediately recognizable as a url).
- Fixed signup button click weirdness.

@jblz 

- Cleaned up url-validation, moved to `/lib`, and added try/catch for parsing. (Thanks for the suggestion!)
- Other issues mentioned in comment responses.

Seems to be some issues with tests... looking into that.

jancavan commented:
In https://github.com/Automattic/wp-calypso/pull/27382/commits/29d1bcefd4a850600c97854879dd3509c9c292a1:

- Removed horizontal scrolling in <660px.
- Added back gradient to browser.
- Made sure escape text has enough right/left padding in <480px.

Other than that, design-wise, LGTM 👍

jonathansadowski commented:
I gave everything another pass. Looks good. Great job! 👍

spen commented on client/signup/steps/import-url/index.jsx:
```diff
@@ -2,171 +2,289 @@
+		const { isLoading, urlInputValue, translate } = this.props;
+		const { showUrlMessage } = this.state;
+		const urlMessage = this.getUrlMessage();
+		const helpPage = 'https://en.support.wordpress.com/import/';
```
These constants could be module variables to avoid assigning on each render, though their proximity to their usage here is a good enough reason to keep them here too, what d'ya think?

spen commented on client/signup/steps/import-url/index.jsx:
```diff
@@ -2,171 +2,289 @@
+						}
+					) }
+					&nbsp;
+					<Button compact onClick={ () => this.exitFlow( '/start' ) }>
```
It's an easy one to miss for sure!
I could have sworn we did have something in place for for the lint warnings in the past - I'll investigate tomorrow :)

spen commented:
Accepting based on code changes - looking really good! Thanks again for those changes! :)
